src/neo4j_graphrag/retrievers/tools_retriever.py

```diff
@@ -0,0 +1,161 @@
+            "would be most helpful in retrieving the relevant information. You can select multiple tools "
+            "if necessary, or none if no tools are appropriate for the query."
+        )
+
+    def get_search_results(
+        self,
+        query_text: str,
+        message_history: Optional[List[LLMMessage]] = None,
+        **kwargs: Any,
+    ) -> RawSearchResult:
+        """Use the LLM to select and execute appropriate tools based on the query.
+
+        Args:
+            query_text (str): The user's query text.
+            message_history (Optional[Union[List[LLMMessage], MessageHistory]], optional):
+                Previous conversation history. Defaults to None.
+            **kwargs (Any): Additional arguments passed to the tool execution.
+
+        Returns:
+            RawSearchResult: The combined results from the executed tools.
+        """
+        if not self._tools:
+            # No tools available, return empty result
+            return RawSearchResult(
+                records=[],
+                metadata={"query": query_text, "error": "No tools available"},
+            )
+
+        try:
+            # Use the LLM to select appropriate tools
+            tool_call_response = self.llm.invoke_with_tools(
+                input=query_text,
+                tools=self._tools,
+                message_history=message_history,
+                system_instruction=self.system_instruction,
+            )
+            # If no tool calls were made, return empty result
+            if not tool_call_response.tool_calls:
+                return RawSearchResult(
+                    records=[],
+                    metadata={
+                        "query": query_text,
+                        "llm_response": tool_call_response.content,
+                        "tools_selected": [],
+                    },
+                )
+
+            # Execute each selected tool and collect results
+            all_records = []
+            tools_selected = []
+
+            for tool_call in tool_call_response.tool_calls:
+                tool_name = tool_call.name
+                tools_selected.append(tool_name)
+
+                # Find the tool by name
+                selected_tool = next(
+                    (tool for tool in self._tools if tool.get_name() == tool_name), None
+                )
+                if selected_tool is not None:
+                    # Extract arguments from the tool call
+                    tool_args = tool_call.arguments or {}
+
+                    # Always include the query_text in the arguments for tools that might need it
+                    tool_args.setdefault("query", query_text)
+
+                    # Execute the tool with the provided arguments
+                    tool_result = selected_tool.execute(**tool_args)
+                    # If the tool result is a RawSearchResult, extract its records
+                    if hasattr(tool_result, "records"):
```

Comment:
We can maybe enforce the fact that `Tool.execute` must return a list of something (or event string?), or do we really want to create fake records?